File: python/legiscore/webhooks.py
"""Verify that a webhook really came from LegiScore.

Getting this wrong is the most common integration bug we see, so the SDK does it:

    from legiscore.webhooks import verify_webhook, InvalidSignature

    @app.post("/webhooks/legiscore")
    def receive(request):
        try:
            event = verify_webhook(request.body, request.headers, secret=WEBHOOK_SECRET)
        except InvalidSignature:
            return 400
        ...

``body`` must be the RAW request bytes. Re-serialising parsed JSON changes the byte string and
every signature check then fails.
"""

from __future__ import annotations

import hashlib
import hmac
import json
import time
from collections.abc import Mapping
from typing import Any

__all__ = ["verify_webhook", "InvalidSignature", "WebhookEvent", "EVENTS"]

SIGNATURE_HEADER = "X-LegiScore-Signature"
TIMESTAMP_HEADER = "X-LegiScore-Timestamp"
EVENT_HEADER = "X-LegiScore-Event"
DELIVERY_HEADER = "X-LegiScore-Delivery"

# The events a case can send. A pause event means the case is waiting on you.
EVENTS = (
    "asset.created",
    "report.auto_triggered",
    "report.started",
    "report.paused.missing_documents",
    "report.paused.review",
    "report.paused.acknowledgements",
    "report.completed",
    "report.failed",
    # Legacy names the older dispatch path still emits.
    "case.paused",
    "case.completed",
)

# Replays older than this are rejected. Five minutes is generous for clock skew and
# still short enough that a captured delivery cannot be resent tomorrow.
DEFAULT_TOLERANCE_SECONDS = 300


class InvalidSignature(Exception):
    """The delivery did not come from LegiScore, or is too old to trust."""


class WebhookEvent(dict[str, Any]):
    """The verified payload, with the delivery metadata pulled up as attributes."""

    def __init__(
        self, payload: dict[str, Any], *, event: str, delivery_id: str, timestamp: int
    ) -> None:
        super().__init__(payload)
        self.event = event
        self.delivery_id = delivery_id
        self.timestamp = timestamp

    @property
    def case_id(self) -> str | None:
        return self.get("case_id") or (self.get("data") or {}).get("case_id")

    @property
    def is_pause(self) -> bool:
        """True when the case has stopped and is waiting for an answer from you."""
        return self.event.startswith("report.paused.") or self.event == "case.paused"


def _header(headers: Mapping[str, Any], name: str) -> str:
    """Case-insensitive lookup — WSGI, ASGI and every framework spell these differently."""
    if name in headers:
        return str(headers[name])
    wanted = name.lower()
    for key, value in headers.items():
        if str(key).lower() == wanted:
            return str(value)
    return ""


def verify_webhook(
    body: bytes | str,
    headers: Mapping[str, Any],
    *,
    secret: str,
    tolerance_seconds: int = DEFAULT_TOLERANCE_SECONDS,
    now: float | None = None,
) -> WebhookEvent:
    """Check the signature and freshness of one delivery, and return its parsed payload.

    Raises :class:`InvalidSignature` if the signature does not match, the timestamp is
    missing or stale, or the body is not the JSON we sent.
    """
    if not secret:
        raise InvalidSignature("No webhook secret configured")

    raw = body.encode("utf-8") if isinstance(body, str) else bytes(body)
    sent_signature = _header(headers, SIGNATURE_HEADER)
    timestamp_header = _header(headers, TIMESTAMP_HEADER)

    if not sent_signature or not timestamp_header:
        raise InvalidSignature("Delivery is missing its signature or timestamp header")

    try:
        timestamp = int(timestamp_header)
    except ValueError:
        raise InvalidSignature(
            f"Timestamp header is not a unix time: {timestamp_header!r}"
        ) from None

    age = abs((time.time() if now is None else now) - timestamp)
    if age > tolerance_seconds:
        raise InvalidSignature(
            f"Delivery is {age:.0f}s old, outside the {tolerance_seconds}s window"
        )

    # Same construction as the sender: HMAC-SHA256 over "<timestamp>." + the raw body,
    # hex-encoded and prefixed with "sha256=".
    expected = (
        "sha256="
        + hmac.new(
            secret.encode("utf-8"), f"{timestamp}.".encode() + raw, hashlib.sha256
        ).hexdigest()
    )
    # Compare as bytes: a WSGI server decodes header bytes as latin-1, so a byte >= 0x80 reaches
    # us as a non-ASCII str and hmac.compare_digest raises TypeError on those. Fail closed with
    # our own exception instead, which is what a partner's `except InvalidSignature` catches.
    if not hmac.compare_digest(
        expected.encode("utf-8"), sent_signature.encode("utf-8", "surrogateescape")
    ):
        raise InvalidSignature("Signature does not match")

    try:
        payload = json.loads(raw.decode("utf-8"))
    except (ValueError, UnicodeDecodeError) as exc:
        raise InvalidSignature(f"Body is not valid JSON: {exc}") from exc
    if not isinstance(payload, dict):
        raise InvalidSignature("Body is not a JSON object")

    return WebhookEvent(
        payload,
        event=_header(headers, EVENT_HEADER) or str(payload.get("event", "")),
        delivery_id=_header(headers, DELIVERY_HEADER) or str(payload.get("delivery_id", "")),
        timestamp=timestamp,
    )

File: node/src/webhooks.ts
/**
 * Verify that a webhook really came from LegiScore.
 *
 * Getting this wrong is the most common integration bug we see, so the SDK does it.
 * `body` must be the RAW request bytes. Re-serialising parsed JSON changes the byte
 * string and every signature check then fails.
 *
 * The HMAC runs on Web Crypto, which every modern runtime has, so this works unchanged in a
 * Next.js Route Handler, on Vercel Edge, in a Cloudflare Worker, in Deno and in Bun.
 * `verifyWebhook` is async because `crypto.subtle` is.
 */

export const SIGNATURE_HEADER = "x-legiscore-signature";
export const TIMESTAMP_HEADER = "x-legiscore-timestamp";
export const EVENT_HEADER = "x-legiscore-event";
export const DELIVERY_HEADER = "x-legiscore-delivery";

/** The events a case can send. A pause event means the case is waiting on you. */
export const EVENTS = [
  "asset.created",
  "report.auto_triggered",
  "report.started",
  "report.paused.missing_documents",
  "report.paused.review",
  "report.paused.acknowledgements",
  "report.completed",
  "report.failed",
  // Legacy names the older dispatch path still emits.
  "case.paused",
  "case.completed",
] as const;

/** Every pause event starts with this, except the legacy name below. */
const PAUSE_EVENT_PREFIX = "report.paused.";
const LEGACY_PAUSE_EVENT = "case.paused";

/**
 * Replays older than this are rejected. Five minutes is generous for clock skew and still
 * short enough that a captured delivery cannot be resent tomorrow.
 */
export const DEFAULT_TOLERANCE_SECONDS = 300;

const MILLISECONDS_PER_SECOND = 1_000;
const HEX_RADIX = 16;
const HEX_DIGITS_PER_BYTE = 2;

/** The delivery did not come from LegiScore, or is too old to trust. */
export class InvalidSignature extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InvalidSignature";
  }
}

export interface WebhookEvent {
  event: string;
  deliveryId: string;
  timestamp: number;
  caseId?: string;
  isPause: boolean;
  payload: Record<string, unknown>;
}

type HeaderSource = Record<string, string | string[] | undefined> | { get(name: string): string | null };

const encoder = new TextEncoder();
const decoder = new TextDecoder();

function header(headers: HeaderSource, name: string): string {
  if (typeof (headers as { get?: unknown }).get === "function") {
    return (headers as { get(n: string): string | null }).get(name) ?? "";
  }
  // Case-insensitive: every framework spells these differently.
  for (const [key, value] of Object.entries(headers as Record<string, string | string[]>)) {
    if (key.toLowerCase() === name) return Array.isArray(value) ? (value[0] ?? "") : String(value ?? "");
  }
  return "";
}

export async function verifyWebhook(
  body: Uint8Array | ArrayBuffer | string,
  headers: HeaderSource,
  options: { secret: string; toleranceSeconds?: number; now?: number },
): Promise<WebhookEvent> {
  const { secret, toleranceSeconds = DEFAULT_TOLERANCE_SECONDS } = options;
  if (!secret) throw new InvalidSignature("No webhook secret configured");

  const raw = toBytes(body);
  const sent = header(headers, SIGNATURE_HEADER);
  const timestampHeader = header(headers, TIMESTAMP_HEADER);
  if (!sent || !timestampHeader) {
    throw new InvalidSignature("Delivery is missing its signature or timestamp header");
  }

  const timestamp = Number(timestampHeader);
  if (!Number.isInteger(timestamp)) {
    throw new InvalidSignature(`Timestamp header is not a unix time: ${timestampHeader}`);
  }

  const nowSeconds = options.now ?? Date.now() / MILLISECONDS_PER_SECOND;
  const age = Math.abs(nowSeconds - timestamp);
  if (age > toleranceSeconds) {
    throw new InvalidSignature(
      `Delivery is ${Math.round(age)}s old, outside the ${toleranceSeconds}s window`,
    );
  }

  // Same construction as the sender: "<timestamp>." + raw body, HMAC-SHA256, hex, sha256= prefixed.
  const digest = await signHmacSha256(secret, concatBytes(encoder.encode(`${timestamp}.`), raw));
  const expected = encoder.encode(`sha256=${digest}`);
  const received = encoder.encode(sent);
  if (!equalsConstantTime(expected, received)) {
    throw new InvalidSignature("Signature does not match");
  }

  let payload: unknown;
  try {
    payload = JSON.parse(decoder.decode(raw));
  } catch (error) {
    throw new InvalidSignature(`Body is not valid JSON: ${String(error)}`);
  }
  if (typeof payload !== "object" || payload === null || Array.isArray(payload)) {
    throw new InvalidSignature("Body is not a JSON object");
  }

  const record = payload as Record<string, unknown>;
  const event = header(headers, EVENT_HEADER) || String(record.event ?? "");
  const data = (record.data ?? {}) as Record<string, unknown>;
  return {
    event,
    deliveryId: header(headers, DELIVERY_HEADER) || String(record.delivery_id ?? ""),
    timestamp,
    caseId: (record.case_id ?? data.case_id) as string | undefined,
    isPause: event.startsWith(PAUSE_EVENT_PREFIX) || event === LEGACY_PAUSE_EVENT,
    payload: record,
  };
}

function toBytes(body: Uint8Array | ArrayBuffer | string): Uint8Array {
  if (typeof body === "string") return encoder.encode(body);
  return body instanceof Uint8Array ? body : new Uint8Array(body);
}

function concatBytes(prefix: Uint8Array, rest: Uint8Array) {
  const joined = new Uint8Array(prefix.length + rest.length);
  joined.set(prefix, 0);
  joined.set(rest, prefix.length);
  return joined;
}

async function signHmacSha256(secret: string, payload: BufferSource): Promise<string> {
  const key = await globalThis.crypto.subtle.importKey(
    "raw",
    encoder.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const signature = await globalThis.crypto.subtle.sign("HMAC", key, payload);
  return toHex(new Uint8Array(signature));
}

function toHex(bytes: Uint8Array): string {
  let hex = "";
  for (const byte of bytes) hex += byte.toString(HEX_RADIX).padStart(HEX_DIGITS_PER_BYTE, "0");
  return hex;
}

/**
 * Compare with no early exit, so the time taken says nothing about how far the two digests
 * matched. Both operands are a fixed-length hex digest, so their length is public and testing
 * it first leaks nothing; the loop then runs the same number of iterations every time.
 */
function equalsConstantTime(expected: Uint8Array, received: Uint8Array): boolean {
  if (expected.length !== received.length) return false;
  let difference = 0;
  for (let index = 0; index < expected.length; index++) {
    difference |= expected[index] ^ received[index];
  }
  return difference === 0;
}
